File: Data Structures & Algorithms/house-robber/submission-16.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        N = len(nums)
        dp = {}
        def dfs(i):
            if i in dp:
                return dp[i]
            if i >= N:
                return 0
            dp[i] =  max(dfs(i+1),nums[i]+dfs(i+2))
            return dp[i]
        return dfs(0)
        # N = len(nums)
        # sum2 = nums[N-1]
        # sum1 = 0
        # for i in range(N-2,-1,-1):
        #     temp = max(nums[i]+sum2,sum1)
        #     sum1 = s
        #     sum2 = sum1
        # return s

File: Data Structures & Algorithms/interleaving-string/submission-16.java
class Solution {
    boolean[][] dp;
    public boolean isInterleave(String s1, String s2, String s3) {
        int M = s1.length(), N = s2.length();
        dp = new boolean[M+1][N+1];
        dp[M][N] = true;
        if(M+N != s3.length())
            return false;
        for(int i=M;i>=0;i--){
            for(int j=N;j>=0;j--){
                if(dp[i][j])
                    continue;
                int k = i+j;
                dp[i][j] = (i<M && s1.charAt(i)==s3.charAt(k)?dp[i+1][j]:false) || 
                    (j<N && s2.charAt(j)==s3.charAt(k)?dp[i][j+1]:false);
            }
        }
        return dp[0][0];
        

        
    }

    // public boolean dfs(int i,int j,String s1,String s2,String s3){
    //     if(i==s1.length() && j == s2.length()){
    //         return true;
    //     }
    //     int k = i+j;
    //     if(dp[i][j]!=null)
    //         return dp[i][j];
    //     if(i<s1.length() && s1.charAt(i)==s3.charAt(k) && dfs(i+1,j,s1,s2,s3)){
    //         dp[i][j] = true;
    //         return true;
    //     } 
    //     if(j<s2.length() && s2.charAt(j)==s3.charAt(k) && dfs(i,j+1,s1,s2,s3)){
    //         dp[i][j] = true;
    //         return true;
    //     }
    //     dp[i][j] = false;
    //     return false;
    // }
}
